piccolo_api/change_password/endpoints.py

```diff
@@ -88,77 +103,96 @@ async def post(self, request: Request) -> Response:
                 raise HTTPException(
                     status_code=401,
-                    detail="Password must be at least 6 characters long.",
+                    detail=error,
                 )
 
         if confirm_password != new_password:
+            error = "Passwords do not match."
+
             if body.get("format") == "html":
                 return self.render_template(
                     request,
-                    template_context={"error": "Passwords do not match."},
+                    min_password_length=min_password_length,
+                    template_context={"error": error},
                 )
             else:
-                raise HTTPException(
-                    status_code=401, detail="Passwords do not match."
-                )
+                raise HTTPException(status_code=401, detail=error)
 
-        if not await user.login(username=user.username, password=old_password):
+        if not await piccolo_user.login(
+            username=piccolo_user.username, password=old_password
+        ):
+            error = "Incorrect password."
             if body.get("format") == "html":
                 return self.render_template(
                     request,
-                    template_context={"error": "Incorrect password."},
+                    min_password_length=min_password_length,
+                    template_context={"error": error},
                 )
-            raise HTTPException(detail="Incorrect password.", status_code=401)
+            raise HTTPException(detail=error, status_code=401)
 
         await piccolo_user.update_password(
             user=request.user.user_id, password=new_password
         )
 
-        # after password changes we invalidate session and redirect user
-        # to login endpoint to login again with new password
-        response = RedirectResponse(
-            url=self._login_url, status_code=HTTP_303_SEE_OTHER
-        )
-        response.delete_cookie("id")
-
         if body.get("format") == "html":
             return self.render_template(
                 request,
                 success=True,
+                min_password_length=min_password_length,
                 login_url=self._login_url,
             )
-        return response
+        else:
+            # After the password changes, we invalidate the session and
+            # redirect the user to the login endpoint.
+            response = RedirectResponse(
+                url=self._login_url, status_code=HTTP_303_SEE_OTHER
+            )
+
+            if self._session_cookie_name:
+                response.delete_cookie(self._session_cookie_name)
+
+            session_table = self._session_table
+            if session_table:
+                # TODO - not sure if deleting or marking inactive is most
+                # appropriate. Requires research.
+                await session_table.delete().where(
+                    session_table.user_id == piccolo_user.id
+                )
+

```

Comment:
I think deleting invalidated sessions is a great and safe idea. Why keep invalidated sessions in db when we force user to log in again and create new session? I think this should also be applied to the logout endpoint.

Reply:
Yes, I was looking at the logout endpoint works, and that just deletes the user's session for their current device.

I was trying to find some concrete recommendations on OWASP for best practices for the following:

 * Marking sessions expired in the database vs deleting them (there's no point keeping old sessions around forever in the database, but it does keep an audit trail).
 * Whether to just delete the session for the current device when logging out / changing password, or for all of that user's devices. My feeling is logging out should be the current device only, but changing password should be all devices.